src/dune_engine/action_builder.ml

```diff
@@ -504,9 +492,121 @@ struct
+  | Map2 (_, a, b) -> can_eval_statically a && can_eval_statically b
+  | All xs -> List.for_all xs ~f:can_eval_statically
+  | Paths_for_rule _ -> false
+  | Paths_glob _ -> false
+  | Deps _ -> true
+  | Dyn_paths b -> can_eval_statically b
+  | Dyn_deps b -> can_eval_statically b
+  | Contents _ -> false
+  | Lines_of _ -> false
+  | Label _ -> true
+  | Or_exn b -> can_eval_statically b
+  | Fail _ -> true
+  | If_file_exists (_, _, _) -> false
+  | Memo _ -> false
+  | Catch (t, _) -> can_eval_statically t
+  | Memo_build _ -> false
+  | Dyn_memo_build _ -> false
+  | Build _ ->
+    (* TODO jeremiedimino: This should be [can_eval_statically b], however it
+       breaks the [Expander.set_artifacts_dynamic] trick that it used to break a
+       cycle. The cycle is as follow:
+
+       - [(rule (deps %{cmo:x}) ..)] requires expanding %{cmo:x}
+
+       - expanding %{cmo:x} requires computing the artifacts DB
+
+       - computing the artifacts DB requires computing the module<->library
+       assignment
+
+       - computhig the above requires knowing the set of source files in a given
+       directory
+
+       - computing the above works by looking at the source tree and adding all
+       targets of user rules
+
+       - computing targets of user rules is done by effectively generating the
+       rules for the user rules, which means interpreting the [(deps
+       %{cmo:...})] thing
+
+       In theory, we should be able to compute the list
+
+       looks odd, if we find a way to break it we should be able to change this. *)
+    false
+
+let static_eval =
+  let rec loop : type a. a t -> unit t -> a * unit t =
+   fun t acc ->
+    match t with
+    | Pure x -> (x, acc)
+    | Map (f, a) ->
+      let x, acc = loop a acc in
+      (f x, acc)
+    | Both (a, b) ->
+      let a, acc = loop a acc in
+      let b, acc = loop b acc in
+      ((a, b), acc)
+    | Seq (a, b) ->
+      let (), acc = loop a acc in
+      let b, acc = loop b acc in
+      (b, acc)
+    | Map2 (f, a, b) ->
+      let a, acc = loop a acc in
+      let b, acc = loop b acc in
+      (f a b, acc)
+    | All xs -> loop_many [] xs acc
+    | Paths_for_rule _ -> assert false
+    | Paths_glob _ -> assert false
+    | Deps _ -> ((), acc >>> t)
+    | Dyn_paths b ->
+      let (x, ps), acc = loop b acc in
+      (x, Deps (Dep.Set.of_files_set ps) >>> acc)
+    | Dyn_deps b ->
+      let (x, deps), acc = loop b acc in
+      (x, Deps deps >>> acc)
+    | Contents _ -> assert false
+    | Lines_of _ -> assert false
+    | Label _ -> ((), acc >>> t)
+    | Or_exn b ->
+      let res, acc = loop b acc in
+      (Result.ok_exn res, acc)
+    | Fail { fail } -> fail ()
+    | If_file_exists (_, _, _) -> assert false
+    | Memo _ -> assert false
+    | Catch (t, v) -> ( try loop t acc with _ -> (v, return ()) )
+    | Memo_build _ -> assert false
+    | Dyn_memo_build _ -> assert false
+    | Build b ->
+      let b, acc = loop b acc in
```

Comment:
This seems to be unreachable code, since `can_eval_statically` returns `false` if `Build` constructor is reachable.

Reply:
Indeed, I removed it